// magealert/alert.js
class MageAlert {
  constructor(timeOut, colors, animation) {
    // Favicon
    this.favicon = document.querySelector("link[rel~='icon']");
    // Taking a copy
    this.lastFavicon = this.favicon ? this.favicon.getAttribute('href') : '';
    // Alert place
    this.place = document.body;
    // Time out
    this.time = timeOut;
    // Default colors
    this.colors = {
      danger: colors.danger || 'red',
      success: colors.success || 'green',
      warning: colors.warning || 'yellow',
      orange: colors.orange || 'orange',
    };
    // Default animation
    this.animation = animation;
  }

  // Change the favicon
  changeFavicon(iconUrl) {
    if (!this.favicon) {
      const favicon = document.createElement('link');
      favicon.rel = 'icon';
      document.getElementsByTagName('head')[0].appendChild(favicon);
    }
    this.favicon.href = iconUrl;
  }

  // Turn all colors including name, rgb, ... to hex
  getHexColorCode(str) {
    var ctx = document.createElement('canvas').getContext('2d');
    ctx.fillStyle = str;
    return ctx.fillStyle;
  }

  // Return a darker or lighter than the input color, level is how lighter or darker
  LightenDarkenColor(color, level) {
    let usePound = false;

    if (color[0] == '#') {
      color = color.slice(1);
      usePound = true;
    }

    let num = parseInt(color, 16);

    let r = (num >> 16) + level;

    if (r > 255) r = 255;
    else if (r < 0) r = 0;

    let b = ((num >> 8) & 0x00ff) + level;

    if (b > 255) b = 255;
    else if (b < 0) b = 0;

    let g = (num & 0x0000ff) + level;

    if (g > 255) g = 255;
    else if (g < 0) g = 0;

    return (usePound ? '#' : '') + (g | (b << 8) | (r << 16)).toString(16);
  }

  // Return a boolian if color is light true else false
  isColorLight(color) {
    const hex = color.replace('#', '');
    const c_r = parseInt(hex.substr(0, 2), 16);
    const c_g = parseInt(hex.substr(2, 2), 16);
    const c_b = parseInt(hex.substr(4, 2), 16);
    const brightness = (c_r * 299 + c_g * 587 + c_b * 114) / 1000;
    return brightness > 155;
  }

  // Create the alert -- color is optional
  makeAlert(alertContainer, progress, text, type, color, animation) {
    type = this.colors[type];
    let bgColor;
    let textColor;

    if (!type && !color) bgColor = 'linear-gradient(135deg, #73a5ff, #5477f5)';

    let isLighter = this.isColorLight(this.getHexColorCode(color || type));

    if (isLighter === false) {
      textColor = '#fff';
    } else {
      textColor = '#333';
    }

    alertContainer.style.border = `rgba(255, 255, 255, 0.3)`;
    alertContainer.style.color = textColor;
    alertContainer.style.background =
      bgColor ||
      `linear-gradient(135deg, ${this.LightenDarkenColor(
        this.getHexColorCode(color || type),
        isLighter ? -100 : 60
      )}, ${color || type})`;

    const svg = document.createElement('svg');
    svg.innerHTML = `<svg style="width: 1rem; cursor: pointer" version="1.1" id="Capa_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
    viewBox="0 0 460.775 460.775" style="enable-background:new 0 0 460.775 460.775;" xml:space="preserve">
    <path d="M285.08,230.397L456.218,59.27c6.076-6.077,6.076-15.911,0-21.986L423.511,4.565c-2.913-2.911-6.866-4.55-10.992-4.55
      c-4.127,0-8.08,1.639-10.993,4.55l-171.138,171.14L59.25,4.565c-2.913-2.911-6.866-4.55-10.993-4.55
      c-4.126,0-8.08,1.639-10.992,4.55L4.558,37.284c-6.077,6.075-6.077,15.909,0,21.986l171.138,171.128L4.575,401.505
      c-6.074,6.077-6.074,15.911,0,21.986l32.709,32.719c2.911,2.911,6.865,4.55,10.992,4.55c4.127,0,8.08-1.639,10.994-4.55
      l171.117-171.12l171.118,171.12c2.913,2.911,6.866,4.55,10.993,4.55c4.128,0,8.081-1.639,10.992-4.55l32.709-32.719
      c6.074-6.075,6.074-15.909,0-21.986L285.08,230.397z"/>
    </svg>`;
    svg.replaceWith(...svg.childNodes);
    svg.style.fill = textColor;
    svg.style.width = '1rem';
    svg.style.cursor = 'pointer';
    const p = document.createElement('p');
    p.innerHTML = text;
    const content = document.createElement('div');
    content.appendChild(p);
    content.appendChild(svg);
    content.style.width = 'fit-content';
    content.style.display = 'flex';
    content.style.flexDirection = 'row nowrap';
    content.style.padding = '0.5rem 1rem';
    content.style.paddingTop = '1rem';
    content.style.justifyContent = 'space-between';
    content.style.alignItems = 'center';
    content.style.gap = '12px';
    content.style.margin = '0 12px';

    alertContainer.appendChild(content);
    alertContainer.appendChild(progress);

    svg.addEventListener('click', (e) => {
      alertContainer.classList.add(`end-${animation || this.animation}`);
      setTimeout(() => {
        if (
          e.target.parentElement.parentElement.parentElement.classList.contains(
            'Alert'
          )
        )
          e.target.parentElement.parentElement.parentElement.remove();
      }, 100);
    });
  }

  // Display and remove alert
  show(text, type, color, animation) {
    // Alert container
    const alertContainer = document.createElement('div');
    alertContainer.className = 'Alert';
    // Changing favicon
    this.changeFavicon(
      'https://drive.google.com/uc?export=view&id=1HSdP6CbcOeNIU53B8yVGV6SPGdAZdNBg'
    );
    // Add progress bar
    const progressContainer = document.createElement('div');
    const progress = document.createElement('div');
    progressContainer.appendChild(progress);
    progressContainer.className = 'progress';
    progressContainer.style.height = '5px';
    progressContainer.style.width = '100%';
    progressContainer.style.backgroundColor = 'transparent';
    progress.style.height = '100%';
    progress.style.backgroundColor = 'rgba(255, 255, 255, 0.3)';
    progress.style.maxWidth = '100%';
    progress.style.transition = 'all 100ms linear';
    progress.style.willChange = 'width';

    if (document.querySelector('.Alert')) {
      let position = document.querySelectorAll('.Alert');
      position = position[position.length - 1].getBoundingClientRect();
      alertContainer.style.top = position.top + 60 + 'px';
    } else {
      alertContainer.style.top = '5%';
    }

    this.makeAlert(
      alertContainer,
      progressContainer,
      text,
      type,
      color,
      animation
    );
    this.styles(animation || this.animation);
    this.place.appendChild(alertContainer);

    // Progress move
    let counter = 0;
    const progressMover = setInterval(() => {
      if (counter >= 99) {
        clearInterval(progressMover);
        setTimeout(() => {
          counter = 0;
        }, 100);
      }
      counter++;
      progress.style.width = `${counter}%`;
    }, this.time / 100);

    setTimeout(() => {
      alertContainer.style.right = '10px';
      setTimeout(() => {
        alertContainer.classList.add(`end-${animation || this.animation}`);
        this.changeFavicon(this.lastFavicon);
        setTimeout(() => {
          alertContainer.remove();
        }, 300);
      }, this.time);
    });
  }

  // Styles static styles for alert card
  styles(animation) {
    let transis = '';
    let anim = '';
    const swing = 'all 0.9s cubic-bezier(.30,0,.15,1.56)';
    const elese = 'transform 0.9s ease';
    const fade = 'opacity 0.9s ease';

    if (animation === 'swing') {
      transis = swing;
    }
    if (animation === 'scale') {
      transis = elese;
      anim = 'scale';
    }
    if (animation === 'fade') {
      transis = fade;
      anim = 'fade';
    }

    let style = document.createElement('style');
    style.type = 'text/css';
    style.innerHTML = `
      .Alert {
        display: flex;
        flex-direction: column;
        justify-content: space-between;
        align-items: center;
        position: fixed;
        right: -100%;
        width: fit-content;
        height: fit-content;
        minWidth: 10rem;
        maxWidth: calc(50% - 21px);
        background-color: rgba(255, 255, 255, 0.3);
        backdrop-filter: blur(10px);
        font-family: sans-serif;
        border-radius: 4px;
        
        animation: 1s alternate linear;
        animation-duration: ${animation === 'bounce' ? '1s' : '0.5s'};
        transition: ${transis};
        animation-name: ${anim};
      }

      .end-scale {
        transform: scale(0);
      }

      .end-swing {
        right: -100% !important;
      }

      .end-fade {
        opacity: 0;
      }

      @keyframes scale {
        0% {
          transform: scale(0);
        }
        50% {
          transform: scale(1.2);
        }
        70% {
          transform: scale(0.9);
        }
        100% {
          transform: scale(1);
        }
      }

      @keyframes fade {
        0% {
          opacity: 0;
        }
        100% {
          opacity: 1;
        }
      }
    `;

    document.getElementsByTagName('head')[0].appendChild(style);
  }
}

/**
 * Removing alert time
 * Change default colors like danger, success, ...
 * Default animation for all alerts
 */
// const alert = new MageAlert(5000, { danger: 'red' }, 'fade');

// document.getElementById('alert-trigger').addEventListener('click', (e) => {
//   e.preventDefault();

//   /**
//    * Alert massage
//    * Default colors like danger, success, ...
//    * Your color.!optional!
//    * Animation type
//    */
//   alert.show('Password is not correct!', 'danger', '', 'fade');
//   // ('#acdfda');
// });

export default MageAlert;
