// @mage-and-co/mage_crop/app.js
import HtmlCss from './html-css.js';
// import MageAlert from 'magealert';
/**
 * main class for doing all the picture cut.
 * @constructor
 */
class MageCrop {
  /**
   * select all dom elements that we need in code and some variables too.
   */
  constructor(language, maxWidth, maxHeight) {
    this.language = language;
    this.maxWidth = maxWidth;
    // Dom ElEMENTs
    this.fileInput = document.getElementById('file');
    this.uploadedFiles = document.getElementById('uploaded');
    this.fileInputContainer = document.getElementById('file-input-container');
    this.apply = document.getElementById('apply');
    this.container = document.getElementById('container');
    this.image = document.getElementById('imageToCrop');
    this.cropperContainer = document.getElementById('cropper-container');
    this.images = document.querySelector('.images');
    this.imageContainer = document.querySelector('.image-container');
    this.saveBtn = document.getElementById('crop-button');
    this.zoom = document.querySelector('.image-size-container');
    this.cropper = null;
    this.events = false;
    this.count = 0;
    this.imagesToCrop;
    // limit height and width of input imagee
    this.limHeight = maxHeight;
    this.limWidth = maxWidth;
    // count event for cropper
    this.isEvent = true;
    // type for rotate
    this.rotate = false;
    this.rotateY = false;
    // indexedDB main image
    this.dbName = 'images';
    this.dbVersion = 1;
    this.dbStoreName = 'images';
    this.store = null;
    this.db = null;
    this.dbData = null;
    this.dbDataEdit = null;
    // indexedDB temp
    this.tempDbName = 'tempImages';
    this.tempDbVersion = 1;
    this.tempDbStoreName = 'tempImages';
    this.tempStore = null;
    this.tempDb = null;
    this.tempDbData = null;
    this.tempDbDataEdit = null;
    this.key = null;

    this.counter = 0;

    // this.alert = new MageAlert(5000, { danger: 'red' }, 'fade');
  }

  /**
   * Alert function will handle all the alert we need in here.
   * @param {string} message The alert message
   * @param {string, number} color the background and text color
   */
  imageAlert(message, color, type) {
    if (type === 'title') {
      // this.alert.show(message, '', color, 'fade');
      const alert = document.createElement('div');
      alert.style.position = 'absolut';
      alert.style.top = '20%';
      alert.style.left = '50%';
      alert.transform = 'translateX(-50%)';
      alert.style.borderColor = color;
      alert.className = 'title-error';

      alert.style.display = 'block';
      alert.style.color = color;
      alert.innerHTML = message;
      if (this.language === 'FA') {
        alert.style.direction = 'rtl';
      } else if (this.language === 'EN') {
        alert.style.direction = 'ltr';
      }

      if (document.querySelector('.title-error')) {
        document.querySelector('.title-error').remove();
      }

      document.querySelector('.image-container').appendChild(alert);

      setTimeout(() => {
        alert.style.borderColor = 'gray';

        alert.remove();
      }, 3000);
    } else {
      this.fileInputContainer.style.borderColor = color;

      this.fileInputContainer.querySelector('.invalid-message').style.display =
        'block';
      this.fileInputContainer.querySelector('.invalid-message').style.color =
        color;
      this.fileInputContainer.querySelector('.invalid-message').innerHTML =
        message;
      if (this.language === 'FA') {
        this.fileInputContainer.querySelector(
          '.invalid-message'
        ).style.direction = 'rtl';
      } else if (this.language === 'EN') {
        this.fileInputContainer.querySelector(
          '.invalid-message'
        ).style.direction = 'ltr';
      }

      setTimeout(() => {
        this.fileInputContainer.style.borderColor = 'gray';

        this.fileInputContainer.querySelector(
          '.invalid-message'
        ).style.display = 'none';
      }, 3000);
    }

    return { message, color };
  }

  /**
   * remove all the active class that use for css effects on tools.
   */
  removeActive() {
    const items = document.querySelectorAll('.tools-item');

    items.forEach((item) => {
      item.classList.remove('active');
    });
  }

  /**
   * the image size might be too bigger than window; this function will fix the image size a little less than the parent container so it always be fit.
   */
  fixImageSize() {
    let aspect = this.image.naturalWidth / this.image.naturalHeight;
    this.image.style.height = '400px';
    this.image.style.width = `${400 * aspect}px`;

    if (400 * aspect >= this.imageContainer.getBoundingClientRect().width) {
      this.image.style.width = `${
        this.imageContainer.getBoundingClientRect().width - 50
      }px`;
      this.image.style.height = `${
        (this.imageContainer.getBoundingClientRect().width - 50) / aspect
      }px`;
    }

    let imageToLeft = this.image.style.width;
    let imageToTop = this.image.style.height;
    imageToLeft = [...imageToLeft];
    imageToTop = [...imageToTop];
    imageToLeft = imageToLeft.filter((item) => {
      if (item === 'p' || item === 'x') {
      } else {
        return item;
      }
    });
    imageToTop = imageToTop.filter((item) => {
      if (item === 'p' || item === 'x') {
      } else {
        return item;
      }
    });
    imageToLeft = parseFloat(imageToLeft.join('')) / 2;
    imageToTop = parseFloat(imageToTop.join('')) / 2;
    let left = this.imageContainer.offsetWidth / 2;
    let top = this.imageContainer.offsetHeight / 2;
    this.image.style.left = left - imageToLeft + 'px';
    this.image.style.top = top - imageToTop + 'px';
  }

  /**
   * check the image rotate after each cut so if it's true function will undo it and if it's false function will do the rotate.
   * x axis.
   * @param {bool} rot
   */
  checkRotate(rot) {
    if (rot) {
      this.cropper.setScale('x', false);
      this.image.classList.remove('rotate');
      this.rotate = false;
      this.image.classList.remove('rotate');
    } else {
      if (!this.rotate) {
        this.cropper.setScale('x', true);
        this.image.classList.add('rotate');
        this.rotate = true;
      } else {
        this.cropper.setScale('x', false);
        this.image.classList.remove('rotate');
        this.rotate = false;
        this.image.classList.remove('rotate');
      }
    }
  }

  /**
   * check the image rotate after each cut so if it's true function will undo it and if it's false function will do the rotate.
   * y axis.
   * @param {bool} rot
   */
  checkRotateY(rot) {
    if (rot) {
      this.cropper.setScale('y', false);
      this.image.classList.remove('rotate-y');
      this.rotateY = false;
      this.image.classList.remove('rotate-y');
    } else {
      if (!this.rotateY) {
        this.cropper.setScale('y', true);
        this.image.classList.add('rotate-y');
        this.rotateY = true;
      } else {
        this.cropper.setScale('y', false);
        this.image.classList.remove('rotate-y');
        this.rotateY = false;
        this.image.classList.remove('rotate-y');
      }
    }
  }

  /**
   * Updating acive image after delete. active will make the image selected.
   * @param {element} element image
   */
  updateImageActive(element) {
    if (element.classList.contains('uploaded-img-container')) {
      let src = element.id;
      const previewImage = document.getElementById('preview-image');
      const title = element.querySelector('.img-title').value;
      const alt = element.querySelector('.img-alt').value;

      if (src) {
        previewImage.setAttribute('src', src);
      } else {
        let imgSrc = document
          .querySelector('.slide-show')
          .querySelector('img')
          .getAttribute('src');
        previewImage.setAttribute('src', imgSrc);
      }
      document.querySelector('.preview-title').value = title;
      document.querySelector('.preview-alt').value = alt;
    } else {
      document.body.classList.remove('show');
    }
  }

  /**
   * display all images stored in indexedDB database.
   * @param {bool} afterDel
   */
  showStoredFiles(afterDel) {
    let data = [];
    let html = '';
    // this.getData();

    if (this.dbData) {
      data = this.dbData;

      html += `
        <div class="mac-btns">
          <button class="confirm" id="confirm">
          </button>
          <button class="minimize" id="minimize">
          </button>
          <button class="close" id="close">
          </button>
        </div>
      `;
      // galery preview image
      html += `
        <div class="preview">
          <img id="preview-image" src="" alt="" />
          <div class="preview-actions">
            <div class="preview-inputs">
              <label for="title">${
                this.language === 'FA' ? ':عنوان عکس' : 'Title'
              }</label>
              <input class="preview-title" type="text" name="title" value="">
              <label for="new-title">${
                this.language === 'FA' ? ':عنوان جایگزین' : 'Alt'
              }</label>
              <input class="preview-alt"  type="text" name="new-title" value="">
            </div>
            <div class="preview-btns edit-delete">
              <button class="edit" id="edit">${
                this.language === 'FA' ? 'برش دوباره' : 'edit'
              }</button>
              <button class="delete">${
                this.language === 'FA' ? 'حذف عکس' : 'Delete'
              }</button>
            </div>
          </div>
        </div>
      `;
      data.forEach((item) => {
        html += `
          <div class="uploaded-img-container" id="${item.imageUrl}">
            <div class="image-big">
              <img class="image-check" src="${item.imageUrl}" alt="${
          item.reTitle
        }" >
              <div class="edit-delete">
                <button class="edit" id="edit">${
                  this.language === 'FA' ? 'برش دوباره' : 'Edite'
                }</button>
                <button class="delete">${
                  this.language === 'FA' ? 'حذف عکس' : 'Delete'
                }</button>
              </div>
            </div>
            <div class="image-inputs">
              <label for="title">${
                this.language === 'FA' ? ':عنوان عکس' : 'Title'
              }</label>
              <input class="img-title" type="text" name="title" value="${
                item.title
              }">
              <label for="new-title">${
                this.language === 'FA' ? ':عنوان جایگزین' : 'Alt'
              }</label>
              <input class="img-alt" type="text" name="new-title" value="${
                item.reTitle
              }">
            </div>
          </div>
        `;
      });

      document;
      this.uploadedFiles.querySelector('.gallery').innerHTML = html;
      this.uploadedFiles.style.direction =
        this.language === 'FA' ? 'ltr' : 'rtl';

      const number = document.querySelectorAll('.uploaded-img-container');
      if (number.length >= 5) {
        document.querySelector('.gallery').style.justifyContent = 'flex-start';
      } else {
        document.querySelector('.gallery').style.justifyContent = 'center';
      }

      if (afterDel) {
        if (document.querySelector('.uploaded-img-container')) {
          const gal = document.querySelector('.gallery');
          setTimeout(() => {
            gal.lastElementChild.classList.add('slide-show');
            this.updateImageActive(gal.lastElementChild);
          }, 300);
        } else {
          document.body.classList.remove('show');
        }
      }
    }
  }

  /**
   * open a dataBase for use indexedDB. in this case it will be 'images';
   */
  openDb() {
    let req = indexedDB.open(this.dbName, this.dbVersion);

    req.onsuccess = (e) => {
      this.db = req.result;
      this.getData();
    };

    req.onerror = function (e) {
      console.error('openDb:', e.target.errorCode);
    };

    req.onupgradeneeded = (e) => {
      this.store = e.currentTarget.result.createObjectStore(this.dbStoreName, {
        keyPath: 'title',
        autoIncrement: false,
      });
    };
  }

  /**
   * open a dataBase for use indexedDB. in this case it will be 'tempImages' and it will be use to store images that wating to cut.
   */
  openTempDb() {
    let req = indexedDB.open(this.tempDbName, this.tempDbVersion);

    req.onsuccess = (e) => {
      this.tempDb = req.result;
    };

    req.onerror = function (e) {
      console.error('openDb:', e.target.errorCode);
    };

    req.onupgradeneeded = (e) => {
      this.tempStore = e.currentTarget.result.createObjectStore(
        this.tempDbStoreName,
        {
          keyPath: 'sort',
          autoIncrement: false,
        }
      );
      this.tempStore.createIndex('title', 'title', { unique: true });
    };
  }

  /**
   * use to create an action. this will open a portal to do actions like delete, add, update and ext....
   * @param {string} store_name
   * @param {string} mode
   */
  getObjectStore(store_name, mode) {
    if (store_name === this.dbName) {
      let tx = this.db.transaction(store_name, mode);
      return tx.objectStore(store_name);
    } else {
      let tx = this.tempDb.transaction(store_name, mode);
      this.tempStore = tx.objectStore(store_name);
      return tx.objectStore(store_name);
    }
  }

  /**
   * store data in indexedDB.
   * @param {object} data the data need to be store
   * @param {bool} end to check if it's the last data
   */
  saveData(data, end) {
    let store = this.getObjectStore(this.dbStoreName, 'readwrite');
    let req;
    try {
      req = store.add(data);
    } catch (e) {
      if (e.name == 'DataCloneError')
        console.log(
          "This engine doesn't know how to clone a Blob, " + 'use Firefox'
        );
      throw e;
    }
    req.onsuccess = (e) => {
      this.showStoredFiles();
      if (end) {
        this.getData();
      }
    };
    req.onerror = (e) => {
      console.log('addPublication error', e.error);
    };
  }

  /**
   * will find the data that match with the data argument from indexedDB.
   * @param {object, string} data key
   */
  getData(data) {
    if (!data) {
      if (typeof store == 'undefined')
        this.store = this.getObjectStore(this.dbStoreName, 'readonly');

      let req;
      req = this.store.count();
      req.onerror = function (e) {
        console.error('add error', this.error);
      };

      req = this.store.openCursor();
      req.onsuccess = (e) => {
        req = this.store.getAll();
        req.onsuccess = (e) => {
          this.dbData = e.target.result;
          this.showStoredFiles();
        };
      };
    } else {
      if (typeof store == 'undefined')
        this.store = this.getObjectStore(this.dbStoreName, 'readwrite');

      let req;
      req = this.store.count();
      req.onerror = function (e) {
        console.error('add error', this.error);
      };

      req = this.store.openCursor();
      req.onsuccess = (e) => {
        req = this.store.get(data);
        req.onsuccess = (e) => {
          this.dbDataEdit = e.target.result;
          this.imageEdit(e.target.result);
        };
      };
    }
  }

  /**
   * will delete the data that match with the key argument from indexedDB.
   * @param {string} key
   */
  deleteData(key) {
    if (typeof store == 'undefined')
      this.store = this.getObjectStore(this.dbStoreName, 'readwrite');

    let req;
    req = this.store.count();
    req.onerror = function (e) {
      console.error('add error', this.error);
    };

    req = this.store.openCursor();
    req.onsuccess = (e) => {
      req = this.store.getAll();
      req.onsuccess = (e) => {
        let deleteReq = this.store.delete(key);
        deleteReq.onsuccess = (evt) => {
          this.getData();
        };
        deleteReq.onerror = function (evt) {
          console.error('deletePublication:', evt.target.errorCode);
        };
      };
    };
  }

  /**
   * take the new data and replace them with the data that already exist in indexedDB.
   * @param {string} imageUrl
   * @param {number, string} key
   * @param {string} title
   * @param {string} reTitle
   */
  updateData(imageUrl, key, title, reTitle) {
    if (typeof store == 'undefined')
      this.store = this.getObjectStore(this.dbStoreName, 'readwrite');

    let req;
    req = this.store.count();
    req.onerror = (e) => {
      console.log(e);
    };

    req = this.store.get(key);
    req.onsuccess = (e) => {
      // Get the old value that we want to update
      let data = e.target.result;

      // update the value(s) in the object that you want to change
      data.imageUrl = imageUrl;
      data.title = title;
      data.reTitle = reTitle;

      // Put this updated object back into the database.
      let requestUpdate = this.store.put(data);
      requestUpdate.onerror = (e) => {
        console.log(e);
      };
      requestUpdate.onsuccess = (e) => {
        if (key !== title) {
          this.store.delete(key);
        }

        this.getData();
      };
    };
  }

  /**
   * save the temp data in indexedDB. this will be the images that wating for cut.
   * @param {object} data the actual data
   * @param {bool} isEdit
   * @param {bool} end
   */
  tempSave(data, isEdit, end) {
    let tempStore = this.getObjectStore(this.tempDbStoreName, 'readwrite');
    let req;
    try {
      req = tempStore.add(data);
    } catch (e) {
      if (e.name == 'DataCloneError')
        console.log(
          "This engine doesn't know how to clone a Blob, " + 'use Firefox'
        );
      throw e;
    }

    req = this.tempStore.openCursor('sort');
    req.onsuccess = (e) => {
      if (end) {
        if (!isEdit) {
          // this.addImage();
          this.getTempImages(false);
        } else {
          // this.editImage();
          this.getTempImages(true);
        }
      }
    };
    req.onerror = (e) => {
      console.log('addPublication error', e.error);
    };
  }

  /**
   * will grab all data we need from temp images dataBase indexedDB.
   * @param {bool} edit to check if it's after an image cut or DOMloading
   */
  getTempImages(edit) {
    if (typeof this.tempStore == 'undefined')
      this.tempStore = this.getObjectStore(this.tempDbStoreName, 'readwrite');

    let req;
    req = this.tempStore.count();
    req.onerror = function (e) {
      console.error('add error', this.error);
    };

    req = this.tempStore.openCursor('sort');

    req.onsuccess = (e) => {
      req = this.tempStore.getAll();
      req.onsuccess = (e) => {
        this.tempDbData = e.target.result;

        if (edit) {
          this.editImage();
        } else {
          this.addImage();
        }
      };
    };
  }

  /**
   * clear all data that stored in temp images dataBase indexedDB.
   */
  clearTempImages() {
    // open a read/write db transaction, ready for clearing the data
    let transaction = this.tempDb.transaction([this.tempDbName], 'readwrite');

    // report on the success of the transaction completing, when everything is done
    transaction.oncomplete = function (e) {};

    // create an object store on the transaction
    let objectStore = transaction.objectStore(this.tempDbName);

    // Make a request to clear all the data out of the object store
    let objectStoreRequest = objectStore.clear();

    objectStoreRequest.onsuccess = function (e) {};
  }

  /**
   * update a data in temp images dataBase indexedDB.
   * @param {string} key
   * @param {object} res the new data
   */
  updateTempImages(key, res) {
    this.tempStore = this.getObjectStore(this.tempDbStoreName, 'readwrite');

    let req;
    req = this.tempStore.count();
    req.onerror = function (e) {
      console.error('add error', this.error);
    };

    req = this.tempStore.openCursor();

    let myIndex = this.tempStore.index('title');
    req = myIndex.get(key);
    req.onsuccess = (e) => {
      let data = e.target.result;

      // update the value(s) in the object that you want to change
      if (data) {
        data.imageUrl = res.imageUrl;
        data.title = res.title;
        data.reTitle = res.reTitle;
        // Put this updated object back into the database.
        let requestUpdate = this.tempStore.put(data);
        requestUpdate.onerror = (e) => {
          console.log(e);
        };
        requestUpdate.onsuccess = (e) => {
          if (key !== res.title) {
            this.tempStore.delete(key);
          }

          this.getTempImages();
        };
      }
    };
  }

  /**
   * will call after a new image added. it will create the DOM element class and ... nedded.
   */
  addImage() {
    this.images.innerHTML = '';
    // remove last crop selector
    if (document.querySelector('.js-crop-grid-holder')) {
      document.querySelector('.js-crop-grid-holder').remove();
    }
    if (document.querySelector('.js-crop-image-holder')) {
      document.querySelector('.js-crop-image-holder').remove();
    }
    const data = this.tempDbData;

    if (data.length > 1) {
      data.forEach((item) => {
        const div = document.createElement('div');
        div.className = 'image-toCrop';
        div.id = item.imageUrl;
        const img = document.createElement('img');
        img.className = 'crop-item';
        img.src = item.imageUrl;
        const inputs = `
          <div>
            <input type="text" name="title" class="title" value="${
              item.title
            }" id="${item.title}" placeholder="${
          this.language === 'FA' ? 'عنوان عکس' : 'Title'
        }">
          </div>
          <div>
            <input type="text" name="re-title" class="re-title" value="${
              item.reTitle
            }" placeholder="${
          this.language === 'FA' ? 'عنوان جایگزین' : 'Alt'
        }">
          </div>
        `;

        const inputsContainer = document.createElement('div');
        inputsContainer.innerHTML = inputs;

        div.appendChild(inputsContainer);
        div.appendChild(img);
        document.getElementById('images').appendChild(div);
      });
      const readyToCrop = document.querySelectorAll('.image-toCrop');

      if (!this.images.querySelector('.active-crop')) {
        if (readyToCrop[this.count]) {
          readyToCrop[this.count].classList.add('active-crop');
        }
      }

      if (this.images.querySelector('.active-crop')) {
        if (document.getElementById('imageToCrop')) {
          this.image.src = this.images.querySelector('.active-crop').id;
        } else {
          this.image.src = data[this.count].imageUrl;
        }
      }
    } else {
      const div = document.createElement('div');
      div.className = 'image-toCrop active-crop';
      div.id = data[0].imageUrl;
      const img = document.createElement('img');
      img.className = 'crop-item';
      img.src = data[0].imageUrl;
      const inputs = `
      <div>
        <input type="text" name="title" class="title" id="${
          data[0].title
        }" placeholder="${this.language === 'FA' ? 'عنوان عکس' : 'Title'}">
      </div>
      <div>
        <input type="text" name="re-title" class="re-title" placeholder="${
          this.language === 'FA' ? 'عنوان جایگزین' : 'Alt'
        }">
      </div>
      `;

      const inputsContainer = document.createElement('div');
      inputsContainer.innerHTML = inputs;

      div.appendChild(inputsContainer);
      div.appendChild(img);
      document.getElementById('images').appendChild(div);
      this.image.src = data[0].imageUrl;
    }

    document.getElementById('crop-edited-button').style.display = 'none';
    document.querySelector('.crop-button').style.display = 'flex';

    this.fixImageSize();
  }

  /**
   * will call after image edited. it will create the DOM element class and replace them with old data.
   */
  editImage() {
    this.images.innerHTML = '';
    // remove last crop selector
    if (document.querySelector('.js-crop-grid-holder')) {
      document.querySelector('.js-crop-grid-holder').remove();
    }
    if (document.querySelector('.js-crop-image-holder')) {
      document.querySelector('.js-crop-image-holder').remove();
    }
    const data = this.tempDbData;

    const div = document.createElement('div');
    div.className = 'image-toCrop active-crop';
    div.id = data[0].imageUrl;
    const img = document.createElement('img');
    img.className = 'crop-item';
    img.src = data[0].imageUrl;
    const inputs = `
    <div>
      <input type="text" name="title" class="title" placeholder="${
        this.language === 'FA' ? 'عنوان عکس' : 'Title'
      }">
    </div>
    <div>
      <input type="text" name="re-title" class="re-title" placeholder="${
        this.language === 'FA' ? 'عنوان جایگزین' : 'Alt'
      }">
    </div>
  `;

    const inputsContainer = document.createElement('div');
    inputsContainer.innerHTML = inputs;

    div.appendChild(inputsContainer);

    div.appendChild(img);
    document.getElementById('images').appendChild(div);
    this.image.src = data[0].imageUrl;

    document.querySelector('.crop-button').style.display = 'none';
    document.getElementById('crop-edited-button').style.display = 'flex';

    let aspect = this.image.naturalWidth / this.image.naturalHeight;
    this.image.style.height = '400px';
    this.image.style.width = `${400 * aspect}px`;
    this.fixImageSize();
    this.removeActive();
    this.rotate = false;
    this.rotateY = false;
  }

  /**
   * will set image preview in gallery state. this is the image that will be large and displayed in gallery as selected image.
   * @param {string} src image Base64 src
   */
  setPreview(src) {
    if (src !== undefined) {
      const previewImage = document.getElementById('preview-image');
      const title = document
        .querySelector('.slide-show')
        .querySelector('.img-title').value;
      const alt = document
        .querySelector('.slide-show')
        .querySelector('.img-alt').value;

      if (src) {
        previewImage.setAttribute('src', src);
      } else {
        let imgSrc = document
          .querySelector('.slide-show')
          .querySelector('img')
          .getAttribute('src');
        previewImage.setAttribute('src', imgSrc);
      }
      document.querySelector('.preview-title').value = title;
      document.querySelector('.preview-alt').value = alt;
    }
  }

  /**
   * remove the selector from previous image that cuted so it won't get duplicat.
   */
  updateImage() {
    // remove last crop selector
    if (
      document.querySelector('.js-crop-grid-holder') ||
      document.querySelector('.js-crop-image-holder')
    ) {
      document.querySelector('.js-crop-grid-holder').remove();
      document.querySelector('.js-crop-image-holder').remove();
    }

    if (this.images.querySelector('.active-crop')) {
      this.image.src = this.images.querySelector('.active-crop').id;
    }
  }

  /**
   * the image to cut is able to move. this function make image ready to move and add the nedded event listeners to it.
   */
  imageMove() {
    let mousePosition;
    let offset = [0, 0];
    let isDown = false;
    const canvas = document
      .querySelector('.js-crop-image-holder')
      .querySelector('canvas');

    // make the image ready
    this.image.style.position = 'absolute';
    canvas.style.position = 'absolute';
    this.image.style.zIndex = '10';
    document.querySelector('canvas').style.zIndex = '-10';
    document.querySelector('.js-crop-grid-holder').style.zIndex = '11';

    const imgMouseDown = (e) => {
      isDown = true;
      offset = [
        this.image.offsetLeft - e.clientX,
        this.image.offsetTop - e.clientY,
      ];
    };
    const imgMouseup = () => {
      isDown = false;
    };
    const imgMouseMove = (e) => {
      e.preventDefault();
      if (isDown) {
        mousePosition = {
          x: e.clientX,
          y: e.clientY,
        };
        this.image.style.left = mousePosition.x + offset[0] + 'px';
        this.image.style.top = mousePosition.y + offset[1] + 'px';
        const cWidth = document.getElementById('crop-result').width;
        const cHeight = document.getElementById('crop-result').height;
        document.getElementById('width').innerHTML = `Width: ${cWidth}px`;
        document.getElementById('height').innerHTML = `Height: ${cHeight}px`;
      }
    };

    // // move image with touch in mobile
    // const touchStart = (e) => {
    //   offset = [
    //     this.image.offsetLeft - e.touches[0].clientX,
    //     this.image.offsetTop - e.touches[0].clientY,
    //   ];
    // };
    // const touchMove = (e) => {
    //   e.preventDefault();

    //   mousePosition = {
    //     x: e.touches[0].clientX,
    //     y: e.touches[0].clientY,
    //   };
    //   this.image.style.left = mousePosition.x + offset[0] + 'px';
    //   this.image.style.top = mousePosition.y + offset[1] + 'px';
    // };

    // Zoom in / out with finger
    let deferent = 0;
    let zoomPinch = 0;
    const pinchStart = (e) => {
      e.preventDefault();

      if (e.touches.length > 1) {
        const finger1 = e.touches[0];
        const finger2 = e.touches[1];
        deferent = finger1.clientX - finger2.clientX;
      }
    };

    const pinchMove = (e) => {
      e.preventDefault();
      if (e.touches.length > 1) {
        const finger1 = e.touches[0];
        const finger2 = e.touches[1];
        const def = finger1.clientX - finger2.clientX;

        const defs = def - deferent;
        zoomPinch = parseFloat(this.image.style.height) + defs;

        this.image.style.height = `${zoomPinch}px`;
        this.image.style.width = `${zoomPinch * aspect}px`;
        this.cropper.updateData();
        this.cropper.drawCroppedImage();

        deferent = def;
      }
    };

    this.image.removeEventListener('mousedown', imgMouseDown, true);
    document.removeEventListener('mouseup', imgMouseup, true);
    document.removeEventListener('mousemove', imgMouseMove, true);
    // this.image.removeEventListener('pointerdown', pinch);

    if (!this.events) {
      this.image.addEventListener('mousedown', imgMouseDown, true);
      document.addEventListener('mouseup', imgMouseup, true);
      document.addEventListener('mousemove', imgMouseMove, true);
      this.image.addEventListener('touchmove', pinchMove);
      this.image.addEventListener('touchstart', pinchStart);
      // this.image.addEventListener('pointerdown', pinch);
      this.events = true;
    }

    // change image size
    let zoomH = this.image.height;
    let aspect = this.image.naturalWidth / this.image.naturalHeight;

    const zoomEvent = (e) => {
      if (
        e.target.id === 'image-pluse' ||
        e.target.classList.contains('image-pluse')
      ) {
        zoomH += 50;
        this.cropper.updateData();
      }
      if (
        e.target.id === 'image-minus' ||
        e.target.classList.contains('image-minus')
      ) {
        zoomH -= 50;
        this.cropper.updateData();
      }

      this.image.style.height = `${zoomH}px`;
      this.image.style.width = `${zoomH * aspect}px`;
      this.cropper.updateData();
      this.cropper.drawCroppedImage();
    };

    const zoomWheel = (e) => {
      zoomH += e.deltaY * -0.1;

      this.image.style.height = `${zoomH}px`;
      this.image.style.width = `${zoomH * aspect}px`;
      this.cropper.updateData();
      this.cropper.drawCroppedImage();
    };

    this.zoom.removeEventListener('click', zoomEvent);
    this.zoom.addEventListener('click', zoomEvent);

    const zoomArea = document.querySelector('.image-container');
    zoomArea.removeEventListener('wheel', zoomWheel);
    zoomArea.addEventListener('wheel', zoomWheel);
  }

  /**
   * in the image cut state we have a list of selected images to cut that we can click on them and switch to that image we want to cut in the moment or back and cut it again
   */
  switchImage() {
    this.images.addEventListener('click', (e) => {
      const divs = document.querySelectorAll('.image-toCrop');
      if (e.target.classList.contains('crop-item')) {
        divs.forEach((div) => {
          div.classList.remove('active-crop');
        });

        divs.forEach((item, index) => {
          item.id === e.target.parentElement.id ? (this.count = index) : false;
        });

        e.target.parentElement.classList.add('active-crop');
        this.updateImage();
        this.fixImageSize();
        this.imageContainer.classList.remove('end');
        this.cropper.updateData();
      }
    });
  }

  /**
   * this will add and remove a class that use for a small animation for live info that shows width, height, x and y in the cut state.
   * @param {bool} is
   */
  infoEfect(is) {
    const info = document.querySelector('.info');
    if (!is) {
      info.classList.add('update');
    } else {
      info.classList.remove('update');
    }
  }

  /**
   * this is the base main function. this will handle cutting image(all the canvas) and selector(the selector to select a part of an image).
   * @param {element} imageElement
   * @param {object} options all the other data and infos nedded in the funciton
   */
  canSet(imageElement, options) {
    const here = this;
    let controller = new AbortController();
    let cropper = {
      key: this,
      // Minimum/maximum boundaries of the crop grid
      minLeft: 0,
      minTop: 0,
      maxLeft: 0,
      maxTop: 0,
      minWidth: 20,
      minHeight: 20,
      maxWidth: 0,
      maxHeight: 0,
      // The crop grid's properties when the mouse down event was triggered
      originalWidth: 0,
      originalHeight: 0,
      originalLeft: 0,
      originalTop: 0,
      // The X and Y coordinates of the mouse down event
      originalMouseX: 0,
      originalMouseY: 0,
      // The X and Y coordinates of the mouse move event
      pageX: 0,
      pageY: 0,
      // The amount the cursor has moved since the mouse was pressed
      deltaX: 0,
      deltaY: 0,
      // The new left and top coordinates of the crop grid
      newLeft: 0,
      newTop: 0,
      // The new width and height of the crop grid
      newWidth: 0,
      newHeight: 0,
      // The element that triggered the mouse down event
      mouseDownElement: null,
      // Page elements
      imageToCrop: null,
      imageOverlay: null,
      gridHolder: null,
      grid: null,
      cropResult: null,
      // The resize handles are stored in an object for easier programmatic access
      resizers: {
        topLeft: null,
        topMid: null,
        topRight: null,
        rightMid: null,
        botRight: null,
        botMid: null,
        botLeft: null,
        leftMid: null,
      },
      // References to the image overlay canvas context, the crop result canvas context,
      // the crop grid style, and the crop result canvas style
      // are kept outside the event handlers for optimal performance
      imageOverlayContext: null,
      cropResultContext: null,
      gridHolderStyle: null,
      cropResultStyle: null,
      // The left, top, right, and bottom coordinates of the source image
      imageLeft: 0,
      imageTop: 0,
      imageRight: 0,
      imageBottom: 0,

      // image size
      imageW: imageElement.naturalWidth / imageElement.width,
      imageH: imageElement.naturalHeight / imageElement.height,
      // Crop grid boundaries
      gridHolderLeft: 0,
      gridHolderTop: 0,
      gridHolderWidth: 0,
      gridHolderHeight: 0,
      // aspect ratios
      aspect: null,
      // is event listener on
      mouseMove: false,
      // mobile touch event
      mobile: false,
      // count for 1 event
      isEvent: this.isEvent,
      // rotate mirror
      mirror: false,
      mirrorY: false,
      // event listeners signals
      // Toggle crop state
      enableCropMode: function (flag) {
        if (flag) {
          // Show the image overlay
          this.imageOverlay.style.removeProperty('opacity');
          // Show the crop grid
          this.gridHolderStyle.visibility = 'visible';
          this.gridHolderStyle.opacity = '1';
        } else {
          this.gridHolderStyle.visibility = 'hidden';
          this.gridHolderStyle.opacity = '0';
        }
        this.updateCropBackground();

        here.imageMove(this.mobile);
        this.drawCroppedImage();
      },
      // update live data
      updateData: function (e) {
        this.imageW = imageElement.naturalWidth / imageElement.width;
        this.imageH = imageElement.naturalHeight / imageElement.height;
      },
      // set mirror for scale
      setScale: function (type, active) {
        if (type === 'x') {
          this.mirror = active;
        }
        if (type === 'y') {
          this.mirrorY = active;
        }
      },
      // Set the output canvas
      setOutputCanvas: function (canvasElement) {
        // Update the output canvas reference
        this.cropResult = canvasElement;
        // Update the output canvas style reference
        this.cropResultStyle = canvasElement.style;
        // Update the output canvas context reference
        this.cropResultContext = canvasElement.getContext('2d');
        // Draw the crop result on to the output canvas
        this.drawCroppedImage();
      },
      // Generate the crop result
      drawCroppedImage: function () {
        const widthToShowPlace = document.getElementById('width');
        const heightToShowPlace = document.getElementById('height');
        const xToShowPlace = document.getElementById('xpos');
        const yToShowPlace = document.getElementById('ypos');

        this.gridHolderLeft =
          this.gridHolder.offsetLeft - here.image.offsetLeft;
        this.gridHolderTop = this.gridHolder.offsetTop - here.image.offsetTop;
        this.gridHolderWidth = this.gridHolder.offsetWidth;
        this.gridHolderHeight = this.gridHolder.offsetHeight;

        // Set the width and height of the output canvas to the width and height of the
        // selected area and adjust the left, top, and bottom margins so that the position
        // of the output image will match the position of the crop grid relative to the source image
        this.cropResult.width = this.gridHolderWidth * this.imageW;
        this.cropResultStyle.marginLeft = this.gridHolderLeft * -1 + 'px';
        this.cropResult.height = this.gridHolderHeight * this.imageH;
        this.cropResultStyle.marginTop = this.gridHolderTop + 'px';
        this.cropResultStyle.marginBottom =
          this.imageToCrop.offsetHeight -
          (this.gridHolderHeight + this.gridHolderTop) +
          'px';
        // Clear the output canvas
        this.cropResultContext.clearRect(
          0,
          0,
          this.gridHolderWidth,
          this.gridHolderHeight
        );
        // show width heihgt x and y
        if (this.cropResult.width !== 0) {
          widthToShowPlace.innerHTML = 'Width: ' + this.cropResult.width + 'px';
          heightToShowPlace.innerHTML =
            'Height: ' + this.cropResult.height + 'px';
        }
        xToShowPlace.innerHTML = 'x: ' + this.gridHolderLeft + 'px';
        yToShowPlace.innerHTML = 'y: ' + this.gridHolderTop + 'px';

        // mirror x
        let rightDis;
        let left;
        let gridWidth;

        // mirror y
        let bottomDis;
        let top;
        let gridHeight;

        // check for scale
        let sx;
        let sy;

        if (this.mirror) {
          rightDis =
            this.gridHolderLeft + this.gridHolderWidth - here.image.width;

          left = rightDis * -1;
          gridWidth = this.gridHolderWidth * -1;

          sx = -1;
        } else {
          left = this.gridHolderLeft;
          gridWidth = this.gridHolderWidth;

          sx = 1;
        }
        if (this.mirrorY) {
          bottomDis =
            this.gridHolderTop + this.gridHolderHeight - here.image.height;

          top = bottomDis * -1;
          gridHeight = this.gridHolderHeight * -1;

          sy = -1;
        } else {
          top = this.gridHolderTop;
          gridHeight = this.gridHolderHeight;

          sy = 1;
        }

        this.cropResultContext.scale(sx, sy);

        // Draw the cropped image
        this.cropResultContext.drawImage(
          this.imageToCrop,
          // Source X and Y coordinates are the crop grid's top and left respectively
          left * this.imageW,
          top * this.imageH,
          // Source width and height are the crop grid's width and height respectively
          this.gridHolderWidth * this.imageW,
          this.gridHolderHeight * this.imageH,
          // The target X and Y coordinates are (0, 0) so that the output image will be
          // drawn at the top left corner of the output canvas; the target width and height
          // are the same as the source width and height respectively
          0,
          0,
          gridWidth * this.imageW,
          gridHeight * this.imageH
        );

        this.updateData();
      },
      // Download the crop result
      downloadCroppedImage: function () {
        // Add comments
        let anchorElement = document.createElement('a');
        let anchorElementStyle = anchorElement.style;
        this.drawCroppedImage();
        anchorElement.href = this.cropResult.toDataURL('image/jpeg');
        // .replace('image/png', 'image/octet-stream');
        let name = document
          .querySelector(`.active-crop`)
          .querySelector('.title').value;

        anchorElement.download = `${this.imageToCrop.src.replace(
          /^.*/,
          `${name}`
        )}`;
        anchorElementStyle.display = 'none';
        anchorElementStyle.visibility = 'hidden';
        anchorElementStyle.opacity = 0;
        document.body.appendChild(anchorElement);
        anchorElement.click();
        setTimeout(() => {
          document.body.removeChild(anchorElement);
          anchorElementStyle = null;
          anchorElement = null;
        });
      },
      // Adjust the boundaries of the crop grid
      // so that it may not spill outside the source image
      fixGrid: function () {
        // my change's for here is change the image with it's container!!!!
        this.gridHolderLeft = this.gridHolder.offsetLeft;
        this.gridHolderTop = this.gridHolder.offsetTop;
        this.gridHolderWidth = this.gridHolder.offsetWidth;
        this.gridHolderHeight = this.gridHolder.offsetHeight;
        this.maxWidth = here.imageContainer.offsetWidth - this.gridHolderLeft;
        this.maxHeight = here.imageContainer.offsetHeight - this.gridHolderTop;
        if (this.mouseDownElement !== this.grid) {
          if (this.gridHolderWidth > this.maxWidth)
            this.gridHolderStyle.width = this.maxWidth + 'px';
          if (this.gridHolderHeight > this.maxHeight)
            this.gridHolderStyle.height = this.maxHeight + 'px';
          if (this.gridHolderWidth < this.minWidth)
            this.gridHolderStyle.width = this.minWidth + 'px';
          if (this.gridHolderHeight < this.minHeight)
            this.gridHolderStyle.height = this.minHeight + 'px';
        } else {
          if (this.gridHolderWidth > this.maxWidth)
            this.gridHolderStyle.left =
              this.gridHolder.offsetLeft -
              (this.gridHolderWidth - this.maxWidth) +
              'px';
          if (this.gridHolderHeight > this.maxHeight)
            this.gridHolderStyle.top =
              this.gridHolder.offsetTop -
              (this.gridHolderHeight - this.maxHeight) +
              'px';
        }
      },
      // Update the image overlay canvas to highlight the selected area of the source image
      updateCropBackground: function () {
        // Fill the image overlay canvas with the default fill colour (#000000)
        this.imageOverlayContext.fillRect(
          0,
          0,
          this.imageOverlay.width,
          this.imageOverlay.height
        );
        // Make the area bound by the crop grid transparent
        this.imageOverlayContext.clearRect(
          this.gridHolder.offsetLeft,
          this.gridHolder.offsetTop,
          this.gridHolder.offsetWidth,
          this.gridHolder.offsetHeight
        );
      },
      // Add comments
      // @params      aspectW = 16
      // @params      aspectH = 9
      setCropRegion: function (aspect) {
        let newH;
        let newW = this.gridHolderWidth;

        if (aspect === '16:9') {
          newH = (newW * 9) / 16;
        }
        if (aspect === '9:16') {
          newH = (newW * 16) / 9;
        }
        if (aspect === '1:1') {
          newH = (newW * 1) / 1;
        }

        this.gridHolderStyle.width = `${newW}px`;

        this.gridHolderStyle.height = `${newH}px`;

        // newwwwwwwwww code
        this.aspect = aspect;

        // Prevent the crop grid from spilling
        this.fixGrid();
        // Update the highlighted area
        this.updateCropBackground();
        // Update the crop result
        this.drawCroppedImage();
      },
      // Resize the crop grid when dragging any of the resize handles
      resizeGrid: function (e) {
        const thiss = this;
        this.updateData();
        // Get the cursor coordinates
        if (this.mobile) {
          this.pageX = e.touches[0].pageX;
          this.pageY = e.touches[0].pageY;
        } else {
          e.preventDefault();
          this.pageX = e.pageX;
          this.pageY = e.pageY;
        }
        // Calculate the amount the cursor has moved since the mouse was pressed
        this.deltaX = this.pageX - this.originalMouseX;
        this.deltaY = this.pageY - this.originalMouseY;
        // Calculate the new left and top positions of the crop grid
        this.newLeft = this.originalLeft + this.deltaX;
        this.newTop = this.originalTop + this.deltaY;
        // The left and top positions should not go outside the source image boundaries
        if (this.newLeft < this.minLeft) {
          this.newLeft = this.minLeft;
          if (
            this.mouseDownElement === this.resizers.botLeft ||
            this.mouseDownElement === this.resizers.leftMid ||
            this.mouseDownElement === this.resizers.topLeft
          )
            this.deltaX = this.newLeft - this.originalLeft;
        } else if (this.newLeft > this.maxLeft) this.newLeft = this.maxLeft;
        if (this.newTop < this.minTop) {
          this.newTop = this.minTop;
          if (
            this.mouseDownElement === this.resizers.topLeft ||
            this.mouseDownElement === this.resizers.topMid ||
            this.mouseDownElement === this.resizers.topRight
          )
            this.deltaY = this.newTop - this.originalTop;
        } else if (this.newTop > this.maxTop) this.newTop = this.maxTop;
        // Initialise the new width and height of the crop grid
        // as the same as the old width and height
        this.newWidth = this.originalWidth;
        this.newHeight = this.originalHeight;
        // Resize the crop grid based on which resize boundary is being dragged
        switch (this.mouseDownElement) {
          // Top left // worked (;)
          case this.resizers.topLeft:
            // Change the top and left positions according to the mouse position
            // and update the width and height so that the right and bottom edges will stay the same
            this.newWidth -= this.deltaX;
            this.newHeight -= this.deltaY;
            if (this.newWidth > this.minWidth) {
              if (thiss.aspect === '16:9') {
                this.gridHolderStyle.top =
                  this.originalTop +
                  this.originalHeight / 2 -
                  this.gridHolderHeight / 2 +
                  'px';
                this.gridHolderStyle.left =
                  this.originalLeft +
                  this.originalWidth -
                  this.gridHolderWidth +
                  'px';
                this.gridHolderStyle.width = (this.newHeight * 16) / 9 + 'px';
              }
              if (thiss.aspect === '9:16') {
                this.gridHolderStyle.top =
                  this.originalTop +
                  this.originalHeight / 2 -
                  this.gridHolderHeight / 2 +
                  'px';
                this.gridHolderStyle.left =
                  this.originalLeft +
                  this.originalWidth -
                  this.gridHolderWidth +
                  'px';
                this.gridHolderStyle.width = (this.newHeight * 9) / 16 + 'px';
              }
              if (thiss.aspect === '1:1') {
                this.gridHolderStyle.top =
                  this.originalTop +
                  this.originalHeight / 2 -
                  this.gridHolderHeight / 2 +
                  'px';
                this.gridHolderStyle.left =
                  this.originalLeft +
                  this.originalWidth -
                  this.gridHolderWidth +
                  'px';
                this.gridHolderStyle.width = (this.newHeight * 1) / 1 + 'px';
              }
              if (thiss.aspect === 'free' || thiss.aspect === null) {
                this.gridHolderStyle.left = this.newLeft + 'px';
                this.gridHolderStyle.width = this.newWidth + 'px';
              }
            }
            if (this.newHeight > this.minHeight) {
              this.gridHolderStyle.top = this.newTop + 'px';
              this.gridHolderStyle.height = this.newHeight + 'px';
            }
            break;
          // Top middle // worked!!(:
          case this.resizers.topMid:
            // Change the top position according to the mouse position
            // and update the height so that the bottom edge will stay the same
            this.newHeight -= this.deltaY;
            if (this.newHeight > this.minHeight) {
              this.gridHolderStyle.top = this.newTop + 'px';
              if (thiss.aspect === '16:9') {
                this.gridHolderStyle.left =
                  this.originalLeft +
                  this.originalWidth / 2 -
                  this.gridHolderWidth / 2 +
                  'px';
                this.gridHolderStyle.width = (this.newHeight * 16) / 9 + 'px';
              }
              if (thiss.aspect === '9:16') {
                this.gridHolderStyle.left =
                  this.originalLeft +
                  this.originalWidth / 2 -
                  this.gridHolderWidth / 2 +
                  'px';
                this.gridHolderStyle.width = (this.newHeight * 9) / 16 + 'px';
              }
              if (thiss.aspect === '1:1') {
                this.gridHolderStyle.left =
                  this.originalLeft +
                  this.originalWidth / 2 -
                  this.gridHolderWidth / 2 +
                  'px';
                this.gridHolderStyle.width = (this.newHeight * 1) / 1 + 'px';
              }
              if (thiss.aspect === 'free' || thiss.aspect === null) {
                this.gridHolderStyle.height = this.newHeight + 'px';
              }
              this.gridHolderStyle.height = this.newHeight + 'px';
            }
            break;
          // Top right //added
          case this.resizers.topRight:
            // Change the width and top position according to the mouse position
            // and update the height so that the bottom edge will stay the same
            this.newWidth += this.deltaX;
            this.newHeight -= this.deltaY;
            if (this.newWidth > this.minWidth) {
              if (thiss.aspect === '16:9') {
                this.gridHolderStyle.width = (this.newHeight * 16) / 9 + 'px';
              }
              if (thiss.aspect === '9:16') {
                this.gridHolderStyle.width = (this.newHeight * 9) / 16 + 'px';
              }
              if (thiss.aspect === '1:1') {
                this.gridHolderStyle.width = (this.newHeight * 1) / 1 + 'px';
              }
              if (thiss.aspect === 'free' || thiss.aspect === null) {
                this.gridHolderStyle.width = this.newWidth + 'px';
              }
            }
            if (this.newHeight > this.minHeight) {
              this.gridHolderStyle.top = this.newTop + 'px';
              this.gridHolderStyle.height = this.newHeight + 'px';
            }
            break;
          // Right middle // worked 0-0
          case this.resizers.rightMid:
            // Change the width according to the mouse position
            this.newWidth += this.deltaX;
            if (this.newWidth > this.minWidth) {
              this.gridHolderStyle.width = this.newWidth + 'px';
              if (thiss.aspect === '16:9') {
                this.gridHolderStyle.top =
                  this.originalTop +
                  this.originalHeight / 2 -
                  this.gridHolderHeight / 2 +
                  'px';
                this.gridHolderStyle.height = (this.newWidth * 9) / 16 + 'px';
              }
              if (thiss.aspect === '9:16') {
                this.gridHolderStyle.top =
                  this.originalTop +
                  this.originalHeight / 2 -
                  this.gridHolderHeight / 2 +
                  'px';
                this.gridHolderStyle.height = (this.newWidth * 16) / 9 + 'px';
              }
              if (thiss.aspect === '1:1') {
                this.gridHolderStyle.top =
                  this.originalTop +
                  this.originalHeight / 2 -
                  this.gridHolderHeight / 2 +
                  'px';
                this.gridHolderStyle.height = (this.newWidth * 1) / 1 + 'px';
              }
              if (thiss.aspect === 'free' || thiss.aspect === null) {
                this.gridHolderStyle.width = this.newWidth + 'px';
              }
            }
            break;
          // Bottom right //added
          case this.resizers.botRight:
            // Change the width and height according to the mouse position
            this.newWidth += this.deltaX;
            this.newHeight += this.deltaY;
            if (this.newWidth > this.minWidth) {
              if (thiss.aspect === '16:9') {
                this.gridHolderStyle.width = (this.newHeight * 16) / 9 + 'px';
              }
              if (thiss.aspect === '9:16') {
                this.gridHolderStyle.width = (this.newHeight * 9) / 16 + 'px';
              }
              if (thiss.aspect === '1:1') {
                this.gridHolderStyle.width = (this.newHeight * 1) / 1 + 'px';
              }
              if (thiss.aspect === 'free' || thiss.aspect === null) {
                this.gridHolderStyle.width = this.newWidth + 'px';
              }
            }
            if (this.newHeight > this.minHeight)
              this.gridHolderStyle.height = this.newHeight + 'px';
            break;
          // Bottom middle // worked 0_0
          case this.resizers.botMid:
            // Change the height according to the mouse position
            this.newHeight += this.deltaY;
            if (this.newHeight > this.minHeight) {
              this.gridHolderStyle.height = this.newHeight + 'px';
              if (thiss.aspect === '16:9') {
                this.gridHolderStyle.left =
                  this.originalLeft +
                  this.originalWidth / 2 -
                  this.gridHolderWidth / 2 +
                  'px';
                this.gridHolderStyle.width = (this.newHeight * 16) / 9 + 'px';
              }
              if (thiss.aspect === '9:16') {
                this.gridHolderStyle.left =
                  this.originalLeft +
                  this.originalWidth / 2 -
                  this.gridHolderWidth / 2 +
                  'px';
                this.gridHolderStyle.width = (this.newHeight * 9) / 16 + 'px';
              }
              if (thiss.aspect === '1:1') {
                this.gridHolderStyle.left =
                  this.originalLeft +
                  this.originalWidth / 2 -
                  this.gridHolderWidth / 2 +
                  'px';
                this.gridHolderStyle.width = (this.newHeight * 1) / 1 + 'px';
              }
              if (thiss.aspect === 'free' || thiss.aspect === null) {
                this.gridHolderStyle.width = this.newWidth + 'px';
              }
            }
            break;
          // Bottom left  Fixed
          case this.resizers.botLeft:
            // Change the height and left position according to the mouse position
            // and update the width so that the right edge will stay the same
            this.newWidth -= this.deltaX;
            this.newHeight += this.deltaY;
            if (this.newWidth > this.minWidth) {
              if (thiss.aspect === '16:9') {
                this.gridHolderStyle.left =
                  this.newWidth -
                  this.originalWidth -
                  this.originalLeft +
                  this.newLeft * 2 +
                  'px';
                this.gridHolderStyle.height = (this.newWidth * 9) / 16 + 'px';
              }
              if (thiss.aspect === '9:16') {
                this.gridHolderStyle.left =
                  this.newWidth -
                  this.originalWidth -
                  this.originalLeft +
                  this.newLeft * 2 +
                  'px';
                this.gridHolderStyle.height = (this.newWidth * 16) / 9 + 'px';
              }
              if (thiss.aspect === '1:1') {
                this.gridHolderStyle.left =
                  this.newWidth -
                  this.originalWidth -
                  this.originalLeft +
                  this.newLeft * 2 +
                  'px';
                this.gridHolderStyle.height = this.newWidth + 'px';
              }
              if (thiss.aspect === 'free' || thiss.aspect === null) {
                this.gridHolderStyle.left = this.newLeft + 'px';
                this.gridHolderStyle.height = this.newHeight + 'px';
              }
            }
            if (this.newWidth > this.minWidth)
              this.gridHolderStyle.width = this.newWidth + 'px';
            break;
          // Left middle
          case this.resizers.leftMid:
            // Change the left position according to the mouse position
            // and update the width so that the right edge will stay the same
            this.newWidth -= this.deltaX;
            if (this.newWidth > this.minWidth) {
              this.gridHolderStyle.left = this.newLeft + 'px';
              this.gridHolderStyle.width = this.newWidth + 'px';
              if (thiss.aspect === '16:9') {
                this.gridHolderStyle.height = (this.newWidth * 9) / 16 + 'px';
                this.gridHolderStyle.top =
                  this.originalTop +
                  this.originalHeight / 2 -
                  this.gridHolderHeight / 2 +
                  'px';
              }
              if (thiss.aspect === '9:16') {
                this.gridHolderStyle.height = (this.newWidth * 16) / 9 + 'px';
              }
              if (thiss.aspect === '1:1') {
                this.gridHolderStyle.height = this.newWidth + 'px';

                this.gridHolderStyle.top =
                  this.originalTop +
                  this.originalHeight / 2 -
                  this.gridHolderHeight / 2 +
                  'px';
                // console.log(this.gridHolderStyle.top);
              }
              if (thiss.aspect === 'free' || thiss.aspect === null) {
                this.gridHolderStyle.width = this.newWidth + 'px';
              }
            }
            break;
          // Crop grid
          case this.grid:
            // Change the left and top positions according to the mouse position
            if (
              this.newLeft < this.gridHolder.offsetLeft ||
              this.gridHolder.offsetWidth < this.maxWidth
            )
              this.gridHolderStyle.left = this.newLeft + 'px';
            if (
              this.newTop < this.gridHolder.offsetTop ||
              this.gridHolder.offsetHeight < this.maxHeight
            )
              this.gridHolderStyle.top = this.newTop + 'px';
            // If the crop grid hits the left/top/right/bottom boundaries, update
            // the original mouse position and the original left position, so that
            // when the user moves the cursor back, the crop grid won't remain locked
            // until the cursor reaches the position where the mouse was pressed
            if (this.pageX >= this.imageLeft && this.pageX <= this.imageRight) {
              if (
                this.gridHolder.offsetLeft === this.minLeft ||
                this.gridHolder.offsetWidth === this.maxWidth
              ) {
                this.originalMouseX += this.deltaX;
                this.originalLeft = this.gridHolder.offsetLeft;
              }
            }
            if (this.pageY >= this.imageTop && this.pageY <= this.imageBottom) {
              if (
                this.gridHolder.offsetTop === this.minTop ||
                this.gridHolder.offsetHeight === this.maxHeight
              ) {
                this.originalMouseY += this.deltaY;
                this.originalTop = this.gridHolder.offsetTop;
              }
            }

            // my codes
            if (thiss.aspect === '16:9') {
              this.gridHolderStyle.top = this.gridHolder.offsetTop;
              this.gridHolderStyle.left = this.gridHolder.offsetLeft;
            }
            if (thiss.aspect === '9:16') {
              this.gridHolderStyle.top = this.newTop + 'px';
              this.gridHolderStyle.left = this.newLeft + 'px';
            }
            if (thiss.aspect === '1:1') {
              this.gridHolderStyle.top = this.newTop + 'px';
              this.gridHolderStyle.left = this.newLeft + 'px';
            }

            break;
          default:
            break;
        }
        // Prevent the crop grid from spilling
        this.fixGrid();
        // Update the highlighted area
        this.updateCropBackground();
        // Update the crop result
        this.drawCroppedImage();
      },
      // Fires when the mouse is released
      endResizingGrid: function (e) {
        // if (!here.mobile) e.preventDefault();
        // Stop listening to the mouse move and mouse up event handlers
        // because they are no longer needed once the mouse is released

        controller.abort();
        document.removeEventListener('mousemove', this.resizeGrid.bind(this));
        document.removeEventListener(
          'mouseup',
          this.endResizingGrid.bind(this)
        );
        document.removeEventListener('touchmove', this.resizeGrid.bind(this));
        document.removeEventListener(
          'touchend',
          this.endResizingGrid.bind(this)
        );
        // No mouse down trigger
        this.mouseDownElement = null;
        // hide info
        here.infoEfect(true);
      },
      // Fires when the mouse is pressed
      startResizingGrid: function (e) {
        controller = new AbortController();
        // The crop grid's initial properties
        this.originalWidth = this.gridHolder.offsetWidth;
        this.originalHeight = this.gridHolder.offsetHeight;
        this.originalLeft = this.gridHolder.offsetLeft;
        this.originalTop = this.gridHolder.offsetTop;
        // The X and Y coordinates of the cursor relative to the page
        if (this.mobile && e.touches) {
          this.originalMouseX = e.touches[0].pageX;
          this.originalMouseY = e.touches[0].pageY;
        } else {
          this.originalMouseX = e.pageX;
          this.originalMouseY = e.pageY;
        }

        // The element that triggered the mouse down event
        this.mouseDownElement = e.currentTarget;
        // Start listening to the mouse move and mouse up events
        // when the mouse is pressed on a resizer element
        // if (!this.mouseMove) {
        document.addEventListener('mousemove', this.resizeGrid.bind(this), {
          signal: controller.signal,
        });
        document.addEventListener('mouseup', this.endResizingGrid.bind(this), {
          signal: controller.signal,
        });
        document.addEventListener('touchmove', this.resizeGrid.bind(this), {
          signal: controller.signal,
        });
        document.addEventListener('touchend', this.endResizingGrid.bind(this), {
          signal: controller.signal,
        });
        // }
        // this.mouseMove = true;
        here.infoEfect(false);
      },
      // Hide the crop grid when it is made transparent so that it may no longer respond to events
      hideGrid: function () {
        if (!this.gridHolderStyle.opacity)
          this.gridHolderStyle.removeProperty('visibility');
      },

      destroy: function () {
        let imageHolder = this.imageToCrop.parentElement;
        imageHolder.parentElement.insertBefore(this.imageToCrop, imageHolder);
        this.gridHolder.removeEventListener(
          'transitionend',
          this.hideGrid.bind(this)
        );
        this.grid.removeEventListener(
          'mousedown',
          this.startResizingGrid.bind(this)
        );
        this.grid.removeEventListener(
          'touchstart',
          this.startResizingGrid.bind(this)
        );
        Object.entries(this.resizers).forEach(
          function ([key, value]) {
            value.removeEventListener(
              'mousedown',
              this.startResizingGrid.bind(this)
            );
            value.removeEventListener(
              'touchstart',
              this.startResizingGrid.bind(this)
            );
          }.bind(this)
        );
        // Remove the style and context references
        this.cropResultStyle = null;
        this.gridHolderStyle = null;
        this.cropResultContext = null;
        this.imageOverlayContext = null;
        // Remove the newly created page elements
        Object.entries(this.resizers).forEach(
          function ([key, value]) {
            value.remove();
          }.bind(this)
        );
        this.grid.remove();
        this.gridHolder.remove();
        this.imageOverlay.remove();
        imageHolder.remove();
        // Remove the page element references
        this.resizers = null;
        this.cropResult = null;
        this.grid = null;
        this.gridHolder = null;
        this.imageOverlay = null;
        this.imageToCrop = null;
        imageHolder = null;
      },
      // Initialise the crop grid and the output canvas
      initialiseGrid: function () {
        let imageToCropClientBoundingRect;
        // The bounding rectangle of the source image
        imageToCropClientBoundingRect =
          this.imageToCrop.getBoundingClientRect();
        // The width and height of the source image
        let imageWidth;
        let imageHeight;

        imageWidth = here.imageContainer.offsetWidth;
        imageHeight = here.imageContainer.offsetHeight;
        // Get the bounding coordinates of the source image
        this.imageLeft = imageToCropClientBoundingRect.left;
        this.imageTop = imageToCropClientBoundingRect.top;
        this.imageRight = imageToCropClientBoundingRect.right;
        this.imageBottom = imageToCropClientBoundingRect.bottom;
        // Set the width and height of the crop grid to the same as those of the source image
        this.imageOverlay.width = imageWidth;
        this.imageOverlay.height = imageHeight;
        // Initialise the image overlay canvas context and the crop grid style references
        this.imageOverlayContext = this.imageOverlay.getContext('2d');
        this.gridHolderStyle = this.gridHolder.style;
        // Set then initial size of the crop grid as 20 pixels smaller than the source image
        this.gridHolderStyle.top = here.image.style.top;
        this.gridHolderStyle.left = here.image.style.left;
        this.gridHolderStyle.width = here.image.style.width;
        this.gridHolderStyle.height = here.image.style.height;
        // Set the maximum width and maximum height of the crop grid to the same as those of the source image
        this.maxWidth = imageWidth;
        this.maxHeight = imageHeight;
        // Set the maximum top and left boundaries of the crop grid
        this.maxLeft = this.maxWidth - this.minWidth;
        this.maxTop = this.maxHeight - this.minHeight;
        // Initialise the output canvas
        this.setOutputCanvas(document.createElement('canvas'));
        // Attach event handlers
        Object.entries(this.resizers).forEach(
          function ([key, value]) {
            value.addEventListener(
              'mousedown',
              this.startResizingGrid.bind(this)
            );
            value.addEventListener(
              'touchstart',
              this.startResizingGrid.bind(this)
            );
          }.bind(this)
        );

        this.grid.removeEventListener(
          'mousedown',
          this.startResizingGrid.bind(this)
        );
        here.image.removeEventListener(
          'mousedown',
          this.startResizingGrid.bind(this)
        );
        this.grid.removeEventListener(
          'touchstart',
          this.startResizingGrid.bind(this)
        );
        here.image.removeEventListener(
          'touchstart',
          this.startResizingGrid.bind(this)
        );
        // Detach event handlers on page unload
        this.grid.addEventListener(
          'mousedown',
          this.startResizingGrid.bind(this)
        );
        // if (this.isEvent) {
        this.grid.addEventListener(
          'touchstart',
          this.startResizingGrid.bind(this)
        );

        here.image.addEventListener(
          'mousedown',
          this.startResizingGrid.bind(this)
        );
        here.image.addEventListener(
          'touchstart',
          this.startResizingGrid.bind(this)
        );

        this.gridHolder.addEventListener(
          'transitionend',
          this.hideGrid.bind(this)
        );
        window.addEventListener('unload', this.destroy.bind(this));
        here.isEvent = false;
        // }
        // Remove the bounding rectangle reference
        imageToCropClientBoundingRect = null;
      },
    };
    let imageHolder = document.createElement('div');
    let resizerClassNames = [
      'top-left',
      'top-mid',
      'top-right',
      'right-mid',
      'bot-right',
      'bot-mid',
      'bot-left',
      'left-mid',
    ];
    let gridTableBody = document.createElement('tbody');
    let addResizer = (value, index, source) => {
      let resizer = document.createElement('div');
      let resizerHandle = document.createElement('div');
      let resizerClassName = `js-crop-resizer ${value}`;
      resizer.className = resizerClassName;
      resizerHandle.className = `${resizerClassName} handle`;
      cropper.resizers[
        value
          .split('-')
          .reduce((x, y) => (x += y[0].toUpperCase() + y.substring(1)))
      ] = resizer;
      cropper.gridHolder.appendChild(resizerHandle);
      cropper.gridHolder.appendChild(resizer);
      resizerHandle = null;
      resizer = null;
    };
    if (
      /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(
        navigator.userAgent
      )
    ) {
      cropper.mobile = true;
    } else {
      cropper.mobile = false;
    }
    cropper.imageToCrop = imageElement;
    cropper.imageOverlay = document.createElement('canvas');
    cropper.gridHolder = document.createElement('div');
    cropper.grid = document.createElement('table');
    cropper.gridHolder.classList.add('js-crop-grid-holder');
    cropper.grid.classList.add('js-crop-grid');
    imageHolder.classList.add('js-crop-image-holder');
    resizerClassNames.forEach(addResizer.bind(this));
    for (let rowLoopIndex = 0; rowLoopIndex <= 2; rowLoopIndex++) {
      let tableRow = document.createElement('tr');
      for (let columnLoopIndex = 0; columnLoopIndex <= 2; columnLoopIndex++)
        tableRow.appendChild(document.createElement('td'));
      gridTableBody.appendChild(tableRow);
      tableRow = null;
    }
    cropper.grid.appendChild(gridTableBody);
    cropper.gridHolder.appendChild(cropper.grid);
    this.imageContainer.appendChild(imageHolder);
    imageHolder.appendChild(imageElement);
    imageHolder.appendChild(cropper.imageOverlay);
    imageHolder.appendChild(cropper.gridHolder);
    cropper.initialiseGrid();
    if (!options) options = {};
    cropper.setOutputCanvas(options.outputCanvas || cropper.cropResult);
    cropper.enableCropMode(!(options.startInCropMode === false));
    gridTableBody = null;
    resizerClassNames = null;
    imageHolder = null;
    return Object.freeze({
      enableCropMode: cropper.enableCropMode.bind(cropper),
      setOutputCanvas: cropper.setOutputCanvas.bind(cropper),
      drawCroppedImage: cropper.drawCroppedImage.bind(cropper),
      downloadCroppedImage: cropper.downloadCroppedImage.bind(cropper),
      setCropRegion: cropper.setCropRegion.bind(cropper),
      destroy: cropper.destroy.bind(cropper),
      endResizingGrid: cropper.endResizingGrid.bind(cropper),
      initialiseGrid: cropper.initialiseGrid.bind(cropper),
      updateData: cropper.updateData.bind(cropper),
      setScale: cropper.setScale.bind(cropper),
    });
  }

  /**
   * this funciton will update just image after cut in the image list those who wating to be cut
   * @param {object} dbData new data
   */
  imageEdit(dbData) {
    this.clearTempImages();
    let currImageBase;
    if (document.querySelector('.slide-show')) {
      currImageBase = document.querySelector('.slide-show');
    }

    let data = dbData;

    const currTitle = data.title;
    const currReTitle = data.reTitle;

    this.tempSave(
      {
        sort: 0,
        reTitle: currReTitle,
        title: currTitle,
        imageUrl: data.imageUrl,
      },
      true,
      true
    );
    // show panel
    document.body.classList.add('crop-state');
    const imageHolder = document.querySelector('.js-crop-image-holder');
    if (imageHolder) {
      imageHolder.remove();
    }

    this.showStoredFiles();
    this.imageEvents();
    this.fixImageSize();
  }

  /**
   * some of event listeners those currently involved with the image to cut.
   */
  imageEvents() {
    const removeSlideShow = () => {
      if (document.querySelector('.slide-show')) {
        let slideShows = document.querySelectorAll('.slide-show');

        slideShows.forEach((item) => {
          item.classList.remove('slide-show');
        });
      }
    };

    const doImageEvent = (e) => {
      const slideShow = document.querySelector('.slide-show');
      const images = document.querySelectorAll('.uploaded-img-container');
      // make the image full screen when clicking on it
      if (e.target.classList.contains('uploaded-img-container')) {
        if (
          !e.target.classList.contains('close') &&
          !e.target.classList.contains('slide-show')
        ) {
          document.body.classList.add('show');
          removeSlideShow();
          e.target.classList.add('slide-show');
          this.setPreview(e.target.querySelector('img').getAttribute('src'));
        }
      }
      if (e.target.classList.contains('image-check')) {
        if (
          !e.target.classList.contains('close') &&
          !e.target.classList.contains('slide-show')
        ) {
          document.body.classList.add('show');
          removeSlideShow();
          e.target.parentElement.parentElement.classList.add('slide-show');
          this.setPreview(e.target.getAttribute('src'));
        }
      }
      if (e.target.classList.contains('image-big')) {
        if (
          !e.target.classList.contains('close') &&
          !e.target.classList.contains('slide-show')
        ) {
          document.body.classList.add('show');
          removeSlideShow();
          e.target.parentElement.classList.add('slide-show');
          setPreview(e.target.getAttribute('src'));
        }
      }
      // delete button
      if (e.target.classList.contains('delete')) {
        let currImageBase;
        let currImage;
        if (document.querySelector('.slide-show')) {
          currImageBase = document.querySelector('.slide-show');
          currImage = currImageBase.querySelector('.img-title').value;
        }

        const imageSrc = currImage;

        this.deleteData(imageSrc);

        images.forEach((item) => item.classList.remove('slide-show'));
        this.uploadedFiles.removeEventListener('click', doImageEvent);
        this.imageEvents();
        this.showStoredFiles(currImageBase.previousElementSibling);
        this.setPreview();
      }
      // edit button
      if (e.target.classList.contains('edit')) {
        let currImageBase;
        let currImage;
        if (document.querySelector('.slide-show')) {
          currImageBase = document.querySelector('.slide-show');
          currImage = currImageBase.querySelector('.img-title').value;
        }
        this.getData(currImage);

        images.forEach((item) => item.classList.remove('slide-show'));
        document.body.classList.remove('show');
        this.uploadedFiles.removeEventListener('click', doImageEvent);
      }
      // remove slide show when click anywhere outside
      if (slideShow) {
        if (
          e.target.classList.contains('close') ||
          e.target.classList.contains('confirm') ||
          e.target.classList.contains('minimize')
        ) {
          images.forEach((item) => item.classList.remove('slide-show'));
          document.body.classList.remove('show');
        }
      }
    };

    this.uploadedFiles.addEventListener('click', doImageEvent);

    // download images
    // document.getElementById('downloadImages').addEventListener('click', (e) => {
    //   e.preventDefault();

    //   this.downloadImages();
    // });
  }

  downloadImages() {
    return this.dbData;
  }

  /**
   * all the other event listners for image, all the tools, ect.
   */
  eventListeners() {
    const here = this;
    this.fileInput.addEventListener('change', (e) => {
      if (e.target.files.length > 4)
        this.images.style.justifyContent = 'flex-start';
      // clear session storage and images ready to crop
      this.clearTempImages();
      this.images.innerHTML = '';
      this.imagesToCrop = [];
      let countFiles = e.target.files;
      for (let i = 0; i < countFiles.length; i++) {
        if (this.fileInput.files[i]) {
          const pattern = /image-*/;
          const file = e.target.files[i];

          if (!file.type.match(pattern)) {
            this.imageAlert(
              `${
                this.language === 'FA'
                  ? 'لطفا فایل با فرمت عکس انتخاب کنید'
                  : 'Please Select File Type Image!'
              }`,
              'red'
            );
            setTimeout(() => {
              this.fileInputContainer.classList.remove('invalid');
            }, 5000);
            return;
          }

          const reader = new FileReader();
          //Read the contents of Image File.
          reader.readAsDataURL(this.fileInput.files[i]);
          reader.onload = function (e) {
            //Initiate the JavaScript Image object.
            const newimage = new Image();

            //Set the Base64 string return from FileReader as source.
            newimage.src = e.target.result;

            //Validate the File Height and Width.
            newimage.onload = function () {
              const height = this.height;
              const width = this.width;
              if (height > this.limHeight || width > this.limWidth) {
                this.imageAlert(
                  `${
                    this.language === 'FA'
                      ? 'سایز عکس بیش از اندازه بزرگ است'
                      : 'The Image Size Is Not Correct'
                  }`,
                  'red'
                );
                return false;
              } else {
                here.counter++;
                let end = false;
                if (here.counter >= here.fileInput.files.length) {
                  end = true;
                  here.counter = 0;
                }

                here.tempSave(
                  {
                    sort: here.counter,
                    reTitle: '',
                    title: here.fileInput.files[here.counter].name,
                    imageUrl: reader.result,
                  },
                  false,
                  end
                );
                document.body.classList.add('crop-state');
                here.fixImageSize();

                const imageHolder = document.querySelector(
                  '.js-crop-image-holder'
                );
                if (imageHolder) {
                  imageHolder.remove();
                }
              }
            };
          };
        }
      }
    });

    this.imageEvents();

    // remove image full screen show
    document.body.addEventListener('click', (e) => {
      const slideShow = document.querySelector('.slide-show');
      const images = document.querySelectorAll('.uploaded-img-container');
      if (slideShow) {
        if (e.target.classList.contains('main-container')) {
          images.forEach((item) => item.classList.remove('slide-show'));
          document.body.classList.remove('show');
        }
      }
    });

    // save button will save croped image in SS
    this.saveBtn.addEventListener('click', (e) => {
      e.preventDefault();
      const currImage = document.querySelector('.active-crop');
      const imageUrl = document
        .getElementById('crop-result')
        .toDataURL('image/jpeg');
      const imageTitle = currImage.querySelector('.title').value;
      const reTitle = currImage.querySelector('.re-title').value;

      let key = currImage.querySelector('.title').id;

      currImage.querySelector('.title').id = imageTitle;

      if (imageTitle === '' || !imageTitle) {
        this.imageAlert(
          `${
            this.language === 'FA'
              ? 'لطفا یک عنوان عکس انتخاب کنید'
              : 'Please add a title'
          }`,
          'red',
          'title'
        );
      } else {
        this.cropper.updateData();
        this.cropper.drawCroppedImage();

        const tempData = this.tempDbData;

        if (tempData.length > 1) {
          if (currImage) {
            this.updateTempImages(key, {
              title: imageTitle,
              reTitle: reTitle,
              imageUrl: imageUrl,
            });
          }

          this.updateImage(currImage);

          this.addImage();

          if (this.images.querySelector('.active-crop').nextSibling) {
            this.images
              .querySelector('.active-crop')
              .nextSibling.classList.add('active-crop');
          }

          this.images
            .querySelector('.active-crop')
            .classList.remove('active-crop');
          this.updateImage(currImage);

          this.fixImageSize();

          // remove mousemove event to prevent make sizes 0
          const el = document.getElementById('crop-result'),
            elClone = el.cloneNode(true);
          el.parentNode.replaceChild(elClone, el);
        } else {
          if (currImage) {
            this.updateTempImages(key, {
              title: imageTitle,
              reTitle: reTitle,
              imageUrl: imageUrl,
            });
          }

          this.updateImage(currImage);

          this.addImage();

          currImage.querySelector('.title').id =
            currImage.querySelector('.title').value;
          currImage.querySelector('.title').value =
            currImage.querySelector('.title').id;
        }

        this.count++;

        let numberOfImages = document.querySelectorAll('.image-toCrop').length;

        if (this.count > numberOfImages - 1) {
          console.log('write what do you want happend after the last image');

          const side = document.querySelector('.side-bar');
          const node = document.getElementById('apply');

          let newApply = node.cloneNode([true]);

          newApply.innerText = `${
            this.language === 'FA' ? 'ذخیره نهایی' : 'Final Confirm'
          }`;
          newApply.classList.add('second-apply');
          newApply.id = 'final-apply';

          if (!document.querySelector('.second-apply')) {
            side.appendChild(newApply);
            document;
            newApply.addEventListener('click', applyHandler);
          }

          // show finish message
          this.imageContainer.classList.add('end');
        }

        this.removeActive();
        this.rotate = false;
        this.rotateY = false;
        this.checkRotate(true);
        this.checkRotateY(true);
      }
    });

    // save edited image
    document
      .querySelector('.crop-edited-button')
      .addEventListener('click', () => {
        const ssItems = this.tempDbData;
        const imageUrl = document
          .getElementById('crop-result')
          .toDataURL('image/jpeg');
        const title = document.querySelector('.title');
        const reTitle = document.querySelector('.re-title');

        let imgTitle;
        let imgRetitle;

        if (title.value !== '') {
          imgTitle = title.value;
        } else {
          imgTitle = ssItems[0].title;
        }
        if (reTitle.value !== '') {
          imgRetitle = reTitle.value;
        } else {
          imgRetitle = ssItems[0].reTitle;
        }
        this.updateData(imageUrl, ssItems[0].title, imgTitle, imgRetitle);

        this.showStoredFiles();

        document.body.classList.remove('crop-state');
        this.fileInput.value = '';
        // remove last crop selector
        if (document.querySelector('.js-crop-grid-holder')) {
          document.querySelector('.js-crop-grid-holder').remove();
        }
        this.imageContainer.classList.remove('end');
        this.removeActive();
        this.rotate = false;
        this.rotateY = false;
        this.checkRotate(true);
        this.checkRotateY(true);
        this.clearTempImages();
      });

    // apply changes
    const applyHandler = (e) => {
      e.preventDefault();
      const newImages = this.tempDbData;
      newImages.forEach((item, index) => {
        let end = false;
        if (index === newImages.length - 1) {
          end = true;
        }
        this.saveData(
          {
            reTitle: item.reTitle,
            title: item.title,
            imageUrl: item.imageUrl,
          },
          end
        );
      });

      document.body.classList.remove('crop-state');
      this.fileInput.value = '';
      // remove last crop selector
      if (document.querySelector('.js-crop-grid-holder')) {
        document.querySelector('.js-crop-grid-holder').remove();
      }
      document.querySelector('html').className = '';
      this.count = 0;
      document.body.classList.remove('crop-state');
      if (document.getElementById('final-apply')) {
        document.getElementById('final-apply').remove();
      }
      here.showStoredFiles();
      this.imageContainer.classList.remove('end');
      this.clearTempImages();
    };
    this.apply.addEventListener('click', applyHandler);

    // cancel cut btn
    document.getElementById('cancel').addEventListener('click', (e) => {
      e.preventDefault();

      // show panel
      document.body.classList.remove('crop-state');
      this.fileInput.value = '';
      this.imageContainer.classList.remove('end');
    });

    // mini mise btn
    document.getElementById('minimize').addEventListener('click', (e) => {
      e.preventDefault();

      // show panel
      document.body.classList.toggle('minimize');
    });
  }

  /**
   * the function that we need to call at first after DOMLoaded.
   */
  init() {
    this.initCropper();
    this.switchImage();
    this.eventListeners();
    this.openDb();
    this.openTempDb();
  }

  /**
   * send data to "initialise" function and make it ready to cut image.
   * some event listeners will be added in here like download image, hide selector, ect.
   */
  initCropper() {
    const here = this;
    // document.addEventListener('DOMContentLoaded', () => {
    document.body.id = this.language;

    document
      .getElementById('imageToCrop')
      .addEventListener('load', function (e) {
        here.cropper = here.canSet(this, {
          outputCanvas: document.getElementById('crop-result'),
        });
      });

    this.removeActive();

    // change crop mode
    document.getElementById('change-mode').addEventListener('click', (e) => {
      if (
        e.target.classList.contains('press') ||
        e.target.parentElement.classList.contains('press')
      ) {
        if (e.target.parentElement.classList.contains('crop-mode')) {
          e.target.parentElement.classList.remove('press');
        } else {
          e.target.classList.remove('press');
        }

        document
          .querySelector('.crop-mode')
          .setAttribute('style', "content: url('../icons/crop-svg.svg')");
      } else {
        if (e.target.parentElement.classList.contains('crop-mode')) {
          e.target.parentElement.classList.add('press');
        } else {
          e.target.classList.add('press');
        }

        document
          .querySelector('.crop-mode')
          .setAttribute('style', "content: url('../icons/move-svg.svg')");
      }
      this.cropper.enableCropMode(!e.target.classList.contains('press'));
    });

    // set Aspect ratio
    document.getElementById('aspect').addEventListener('click', (e) => {
      this.cropper.setCropRegion('free');
      this.removeActive();
      if (e.target.tagName === 'IMG') {
        e.target.parentElement.classList.add('active');
      } else {
        e.target.classList.add('active');
      }
    });
    document.getElementById('aspect1').addEventListener('click', (e) => {
      this.cropper.setCropRegion('16:9');
      this.removeActive();
      if (e.target.tagName === 'IMG') {
        e.target.parentElement.classList.add('active');
      } else {
        e.target.classList.add('active');
      }
    });
    document.getElementById('aspect2').addEventListener('click', (e) => {
      this.cropper.setCropRegion('9:16');
      this.removeActive();
      if (e.target.tagName === 'IMG') {
        e.target.parentElement.classList.add('active');
      } else {
        e.target.classList.add('active');
      }
    });
    document.getElementById('aspect3').addEventListener('click', (e) => {
      this.cropper.setCropRegion('1:1');
      this.removeActive();
      if (e.target.tagName === 'IMG') {
        e.target.parentElement.classList.add('active');
      } else {
        e.target.classList.add('active');
      }
    });

    // mirror
    document.getElementById('mirror').addEventListener('click', (e) => {
      here.checkRotate();
      here.cropper.updateData();
      here.cropper.drawCroppedImage();
    });

    // mirror Y
    document.getElementById('mirror-y').addEventListener('click', (e) => {
      here.checkRotateY();
      here.cropper.updateData();
      here.cropper.drawCroppedImage();
    });

    // click and download selected image
    document
      .getElementById('download-button')
      .addEventListener('click', (e) => {
        this.cropper.downloadCroppedImage(e);
      });

    // check the image side and scale if needed in preview
    const images = document.querySelectorAll('.image-check');
    images.forEach((item) => {
      if (parseInt(item.naturalHeight) > 200) {
        item.parentElement.classList.add('scale');
      }
    });
    // });
  }
}

class Cutter {
  constructor(containerElement, options) {
    this.containerElement = containerElement;
    this.language = options.language;
    this.maxWidth = options.maxWidth;
    this.maxHeight = options.maxHeight;
    this.cropperRef = null;
  }

  init() {
    const htmlCss = new HtmlCss(this.containerElement, this.language);
    htmlCss.init();

    const cropper = new MageCrop(this.language, this.maxWidth, this.maxHeight);
    cropper.init();
    this.cropperRef = cropper;
  }

  getData() {
    return this.cropperRef.downloadImages();
  }
}

// const cropper = new Cutter(document.body, {
//   language: 'EN',
//   maxWidth: 6000,
//   maxHeight: 6000,
// });

// cropper.init();

export default Cutter;
