// @mage-and-co/mage_crop/html-css.js
class HtmlCss {
  constructor(wrapper, language) {
    this.language = language;
    // the container wrapper !can be body
    this.wrapper = wrapper;
    // @ cropperContainre
    this.cropperContainer = this.create(
      'div',
      'page-container',
      'cropper-container'
    );
    // @ viewContainer
    this.viewContainer = this.create('div', 'main-container');
  }
  // all html nedded
  html() {
    this.inner(this.cropperContainer, 'cropper');
    this.inner(this.viewContainer, 'view');

    this.insert(this.wrapper, this.cropperContainer);
    this.insert(this.wrapper, this.viewContainer);
  }
  // html tags creator
  create(tag, className, id) {
    const htmlTag = document.createElement(tag);
    if (id) htmlTag.id = id;
    if (className) htmlTag.className = className;
    return htmlTag;
  }
  // append child
  insert(container, child) {
    container.appendChild(child);
  }
  // add inner html
  inner(element, target) {
    if (target === 'cropper') {
      element.innerHTML = `
      <div class="image-side-container">
        <!-- tools container -->
        <div class="tools">
          <!-- zoom buttons -->
          <div class="image-size-container">
            <button id="image-pluse" class="image-change-size"><img class="image-pluse" src=""
                alt="+"></button>
            <button id="image-minus" class="image-change-size"><img class="image-minus" src=""
                alt="-"></button>
          </div>

          <!-- set aspect ratio -->
          <button id="aspect" class="tools-item">free</button>
          <button id="aspect1" class="tools-item">16:9</button>
          <!-- set aspect ratio -->
          <button id="aspect2" class="tools-item">9:16</button>
          <!-- set aspect ratio -->
          <button id="aspect3" class="tools-item">1:1</button>

          <!-- hide cut selector -->
          <button class="crop-mode-btn" id="change-mode"><img class="crop-mode" src=""
              alt="crop"></button>

          <!-- Download the cropped image -->
          <button id="download-button"><img src="" alt="Download" class="download-cropped-image"></button>

          <!-- mirror image -->
          <button id="mirror"><img src="" alt="mirror-x" class="mirror-x"></button>

          <button id="mirror-y"><img src="" alt="mirror-y" class="mirror-y"></button>

          <!-- save button -->
          <button id="crop-edited-button" class="crop-edited-button">ذخیره</button>
          <button id="crop-button" class="crop-button">${
            this.language === 'FA' ? ' تایید' : 'Next'
          }</button>
        </div>

        <!-- Wrapper DIV that will display its contents as an inline block -->
        <div class="image-container">
          <!-- The image to be cropped -->
          <img id="imageToCrop" src="" alt="Image" />
        </div>

        <!-- Toggle button to enter/leave crop mode -->
        <aside class="side-bar">
          <div class="images" id="images"></div>
        </aside>

        <div class="mac-btns">
          <!-- aply change button -->
          <button class="apply-btn" id="apply"></button>
          <!-- yellow button -->
          <button class="yellow-btn" id="minimize"></button>
          <!-- aply change button -->
          <button class="cancel-btn" id="cancel"></button>
        </div>
      </div>

      <!-- Wrapper DIV that will display its contents as an inline block -->
      <div class="canvas-container">
        <!-- The output canvas -->
        <canvas id="crop-result" width="200" height="300"></canvas>
      </div>

      <!-- show width, height, x, y live -->
      <div class="info">
        <span class="info-item" id="width" placeholder="width"></span>
        <span class="info-item" id="height" placeholder="height"></span>
        <span class="info-item" id="xpos" placeholder="x"></span>
        <span class="info-item" id="ypos" placeholder="y"></span>
      </div>`;
    } else if (target === 'view') {
      element.innerHTML = `
      <div class="whole-container">
        <div class="title" style ="direction: ${
          this.language === 'FA' ? 'ltr' : 'rtl'
        }">${this.language === 'FA' ? 'عکس ها' : 'Pictures'}
        <p class="title-guide">${
          this.language === 'FA'
            ? 'عکس هارا انتخل یا در باکس رها کنید'
            : 'Choose Pictures Or Drag And Drop Them In Box.'
        }</p>
        </div>
        <div class="content">
          <form action="" class="file-input-container" id="file-input-container">
            <label for="file" class="file-input">
            ${
              this.language === 'FA'
                ? 'فایل هارا انتخاب کنید'
                : 'Choose Pictures'
            }
              <input type="file" name="file" id="file" draggable="true" multiple accept="image/*">
            </label>
            <p class="invalid-message"></p>
          </form>
          <div class="uploaded" id="uploaded">
            <div class="gallery"></div>
          </div>
        </div>
      </div>`;
    }
  }
  // all css nedded
  css() {
    const css = document.createElement('link');
    css.rel = 'stylesheet';
    css.href = './css/style.css';
    document.head.appendChild(css);
  }
  // run and create
  init() {
    // this.css();
    this.html();
  }
}

export default HtmlCss;
